### ISSUE TITLE
Update data for Safari 9.1/OS X with new criteria

### ISSUE THREAD

**Issue by dstorey** [stevefaulkner/HTML5accessibility#43]
I can likely run this test as I have a Mac. Will leave data here for someone else to double check rather than directly PR as I work for a different browser vendor.

**Comment by dstorey** [comment]
My Safari 9.1 results for first two sections. I'll test rest tomorrow. Mostly same as already in the results page as I had started filling it out earlier.

Group/Sectioning

**Article**: No change from current

Accessibly supported: Yes
Implemented: Yes
Mapped: Yes
Name/Desc: Yes

**Section**: 

Accessibly Supported: partial*?
Implemented: Yes
Mapped: Partial?*
Name/Desc: Yes

*Role and RoleDescription are right but spec says AXSubrole: **AXLandmarkRegion**, but it uses **AXDocumentRegion**. Either spec or browser should be fixed. It is inconsistent with others, which do use AXLandmarkWhatever. This is same if it has accessible name or not so isn't doing anything fancy.

**Navigation**: No change

Accessibly supported: Yes
Implemented: Yes
Mapped: Yes
Name/Desc: Yes

**Aside**: No change

Accessibly supported: Yes
Implemented: Yes
Mapped: Yes
Name/Desc: Yes

**Header**: No change

Accessibly supported: Yes
Implemented: Yes
Mapped: Yes
Name/Desc: Yes

**Footer**: No change

Accessibly supported: Yes
Implemented: Yes
Mapped: Yes*
Name/Desc: Yes

*Spec says AXRoleDescription: 'content'. Actually uses 'footer'. I believe spec should change. This is consistent with Edge too I believe.

**Figure**:

Accessibly supported: Partial
Implemented: Yes
Mapped: No
Name/Desc: Partial. Figcaption not used

**Figcaption**:

Accessibly supported: Yes
Implemented: Yes
Mapped: Yes.

Technically this is a yes, but current test fails Safari due to not using Figcaption in Figure's name/desc. Should this be part of the criteria?

Also the spec for mapping in Safari seems like the default values: AXGroup/nil/group, so it is possible there is no specific handling. A11y mappings for other APIs are more expressive. (e.g. says it is text for UIA, etc.) Should spec be  more specific for AX here?

**Main**: No change

Accessibly supported: Yes
Implemented: Yes
Mapped: Yes
Name/Desc: Yes

-----

Graphics/Media

**Canvas**

Accessibly supported: Partial
Implemented: Yes.
ShowFocusIfNeeded: Yes
Mapped: No*
SubDOM: Yes. Elements are in tree and voiceover reads out.

*Spec says AXImage/(nil)/"image". Actually AXGroup/(nil)/group

**Audio**

Accessibly supported: Partial
Implemented: Yes.
Mapped: Yes*
Keyboard: No (VO only)

* Mapping is more specific than spec, so that should probably be updated. AXGroup/AXAudio/"audio playback", rather than AXGroup/(nil)/"group".

**video**

Accessibly supported: Partial
Implemented: Yes.
Mapped: Yes*
Keyboard: No (VO only)

* Mapping is more specific than spec, so that should probably be updated. AXGroup/AXVido/"video playback", rather than AXGroup/(nil)/"group".

**Track**

Accessibly supported: Yes
Implemented: Yes
Subtitles: Yes (WebVTT only)

Notes:

aria-label mapped as accessibilityLabel. Spec says to use AXDescription. Assume same thing?
aria-labelledby if references one element (like in tests), suppose to map to AXTitleUIElement. Actually maps to accessibilityLabel. I give these a pass even though different to spec, as it seems right.

**Comment by dstorey** [comment]
Controls

**Search**

Accessibly supported: Partial
Implemented: Yes
Mapped: Yes*
Name/Desc: Partial*


* Technically it is wrong per spec as it says to map as vanilla text box, but I think that is wrong. Safari actually maps it to have the search field semantics: AXTextField/AXSearchField/"search text field"
* Label (wrapped or with id/for) is never used as accessible name. Neither AXTitle (for visual text) AXDescription (for invisible), AXTitleUIElement (whatever that is for, to AXLabel (which in reality is usually used in Safari but not mentioned in spec). Perhaps this is correct as there is a visual label, but mappings specs says it should be accessible name and it isn't.

**Tel **

Accessibly supported: Partial
Implemented: Yes
Mapped: No*
Name/Desc: Partial*

* Maps as a regular HTML4 input=text, which is correct per spec, but not intention, and less expressive than Search in Safari and the mappings for other a11y APIs. Existing HTML5A11y also dings Safari for not mapping as a Telephone field, so I recommend this being "no" and updating spec so that AX is as expressive as UIA, IA2, etc. If it is decided this is yes, then search should be No. 
* Same issue as search for name/desc

**URL **

Accessibly supported: Partial
Implemented: Yes
Mapped: No*
Name/Desc: Partial*
Error state: No

* Both same as issues for Tel
* doesn't expose any errors at all

**Email **

Accessibly supported: Partial
Implemented: Yes
Mapped: No*
Name/Desc: Partial*
Error state: No

* All same as Email.

**Datetime-local**

N/A

**Date**

N/A

**Month**

N/A

**Week**

N/A

**Time**

N/A

**Number**

Accessibly supported: Partial
Implemented: Yes
Mapped: ?* 
Name/Desc: Partial*
Keyboard: Yes*

* The buttons inside number are mapped like spec, but not text field itself (generic html4 style text field). No indication that it is for numbers with a subrole or role description (same issue as other fields.) Is spec wrong and it should be the buttons not the field that is mapped as an incrementor?
* label is never used as accessible name (title is correctly accessible description when there is a label however)
* Can't tab to buttons but up/down spins the value. I'm considering this a pass, but I guess that is up for debate?


**Range**

Accessibly supported: Partial
Implemented: Yes
Mapped: Yes 
Name/Desc: Partial*
Keyboard; Yes

* label is never used as accessible name (title is correctly accessible description when there is a label however)

**Color**

N/A

**Datalist**

N/A

**Output**

Accessibly supported: Partial
Implemented: Yes
Mapped: No* 
Name/Desc: Partial*

* Doesn't show in tree until a value is added. Once a value is added it is AXStaticText just like normal text nodes, or AXGroup if it has an aria attribute.
*  Like inputs, label is never used as accessible name. titles are never used as accessible name or description. Basically not treated like a form element.

**Progress**

Accessibly supported: Partial
Implemented: Yes
Mapped: Partial* 
Name/Desc: Partial*

* Element mapped correctly, but min and max attributes are always 0 for AXMinValue and AXMacValue.
* Like inputs, label is never used as accessible name (title is correctly accessible description when there is a label however)

**Meter**

Accessibly supported: Partial
Implemented: Yes
Mapped: Partial* (exactly same mapping as Progress)
Name/Desc: Partial*

* Element mapped correctly, but min and max attributes are always 0 for AXMinValue and AXMacValue. AX doesn't have a way of mapping the goodness of the value like UIA does. Should this be added to spec? It seems to be something that adds value (visual users see red/amber/green).
* Like inputs, label is never used as accessible name (title is correctly accessible description when there is a label however)


**Details**

Accessibly supported: Yes
Implemented: Yes
Mapped: Yes*
Keyboard: Yes

* Mapping spec has generic AXGroup/nil/group. In reality it is AXGroup/AXDetails/group. Spec should probably be updated. The AXRoleDescription doesn't look consistent with others that have a custom AXSubrole though, so maybe it is partial? Expanded state is correct.

**Summary**

Accessibly supported: Partial
Implemented: Yes
Mapped: No*
Name/Desc: Partial

* Mapping is AXDisclosureTriangle/nil/"disclosure triangle" and other APIs expose that it is some form of button that can be expanded. Safari uses AXGroup/AXSummary/group, which is consistent with Details, but seems to give no indication about being an actionable element. It could be that this is correct, in which case spec should be updated.
* Summary subtree not used as accessible name. title not used as accessible desc when summary is non-empty, Doesn't create an accessible name when summary is empty and no other accessible name.


**Toolbar**

N/A

**Context**

N/A

**Menuitem**

N/A

**Dialog**

N/A

-----

Propertries

**hidden**:

Accessibly supported: Yes
Implemented: Yes
Mapped: Yes

** required**:

Accessibly supported: Partial (assuming mapped is yes or partial)
Implemented: Yes
Mapped: Yes*?
Error: No*

* I can't find AXRequired in Xcode's Accessibility Inspector. It isn't listed in the attributes. It does show in the dev tools (Node -> Accessibility) as Required: Yes, but no idea if that is getting that from AXRequired or if it is just detecting it has a required HTML attribute. The current HTML5acessibility gives Safari a partial for not mapping AXinvalid. I also can't find that. If this is something that should be included in criteria and Safari still doesn't map then it mapped should be partial (assuming AXRequired maps correctly)
* Error is no as it doesn't even present an error, never mind expose to screen readers, so there is no way to know it is in error state.

**Placeholder**:

Accessibly supported: Partial
Mapped: Yes
Name/Desc: No*

* Placeholder never added in any of the various AX name and description properties, just the AXPlaceholderValue property.

**Comment by dstorey** [comment]
If assumptions are all correct, and we give partial for anything that has at least one partial (meaning Safari has no No's now) the score will be up to 68%

X = 100 /  (40 - 12) = 3.5714 
Total  = 100 - (0 * 3.5714) -  (18 * .5 * 3.5714) = 67.8574

**Comment by dstorey** [comment]
If figcaption is partial instead (current scoring where figcaption needs to be in naming for figure) the score would be 66%. 

If criteria is there has to be one sub-criteria with a yes, or all sub-criteria to be partial to get partial, I think (depending on the unclear results), the score would be 54%. This kind of grading would but much harder to understand, but it would stop for example Edge (pre 14) and IE getting partials just because they support some of the naming/desc algorithm.

**Comment by stevefaulkner** [comment]
@dstorey i get 76%: reason being that at some which got a partial for name/desc catually pass its just not clear from the object inspection tools that they do provide correct accname as OSX has some funky methods for exposing. This needs to be better documented in the specs.
X = 100 / (40 - 12) = 3.714
Total = 100 - (0 * 3.714) - (13* .5 * 3.714) =  75.86

I went with the simple calc of: if feature has one (acc related) partial then its an overall partial. Note I didn't count the "Error state mapped" fail towards the headline score for a feature as i think it should treated be a separate feature.

**Comment by dstorey** [comment]
@stevefaulkner I consider not exposing the error state a fail (like in all other browsers, as no one gets it right) as it is a major a11y issue for each of the controls. Safari goes a step further by not even showing the error messages to sighted users either (in reality they probably just did minimal implementation so iOS Safari gets the special keyboards). Note in the spec it doesn't prescribe how the errors are mapped, so I'm considering whether they're exposed in a way a screen reader can get to them (we're using controllerFor etc. in Edge14 implementation to fix this issue)

**Comment by stevefaulkner** [comment]
@dstorey not saying it isn't an issue, but that it is not a core feature of a particular control, it is only invoked when validation is in play, which is why i suggested it should be treated as a separate feature. But if you do see it as integral to the control, then lets go with it.

**Comment by dstorey** [comment]
@stevefaulkner I'm leaning towards yes, because in general the main benefit these HTML5 controls have  (beyond the keyboard on mobile) is that you get free validation without having to write your own JS. But if the errors are not exposed in a way a SR can consume them you can't use them if you care about a11y

**Comment by dstorey** [comment]
@stevefaulkner I think the final score for this should be around 73% including the error exposure fails. This includes 15 partials. I didn't include section as partial as you didn't but it is different to spec and the other landmarks. Should the spec be changed here?

I included number as partial as it fails the error one. You list it as full on the site. I also counted placeholder as partial due to not including it in name/desc calc. This is inconsistent with other browsers and doesn't match the spec. IT is probably why Safari has funky behaviour where it always reads the placeholder first, which to me is weird.

X = 100 /  (40 - 12) = 3.5714 
Total  = 100 - (0 * 3.5714) -  (15 * .5 * 3.5714)  = 73.2145
